### ISSUE TITLE
Navigating grid rows with up/down keyboard arrow keys

### ISSUE THREAD

**Issue by alexandery** [angular-ui/ui-grid#2362]
Unless I'm missing something, I don't seem to be able to navigate rows in the grid using my keyboard. Is this supported? If so, what do I need to set to enable the functionality?

**Comment by longshot5112** [comment]
@nippur72 Where did you define rowNav?

**Comment by nippur72** [comment]
@longshot5112 for my test I hardcoded it manually e.g.`var rowNav = true;`

**Comment by smithscripts** [comment]
Has anyone had any luck implementing a solution for this yet. @nippur72 I've tried you suggestion and it seems to mostly work, but when you get to the bottom row the scrolling does not align with the row that is selected. Especially if you are holding down the down arrow key.

**Comment by alexandery** [comment]
@smithscripts Sorry, I sort of dropped this. Seemed like ui-grid guys were not interested in building this out and I simply didn't have time to invest into building it out myself. I'd love to have it, but it's a nice-to-have for me at this moment.

**Comment by Chyzwar** [comment]
I would like to see this. UI Grid is perfect for me but I need to have elegant way to map rows to underlining models. In my case at least row navigation would be better because row in grid equal to model instance. 

I will try maybe to implement this and do PR but I am struggling at Uni and my day-job is pushing for deadlines.

**Comment by gokhanoner** [comment]
Hi,

You can achieve this by doing below:

```
$scope.gridApi.cellNav.on.navigate($scope,function(newRowCol, oldRowCol){
  $scope.gridApi.selection.selectRow(newRowCol.row.entity);
})
```

**Comment by anticommander** [comment]
Thank you @gokhanoner! Works perfectly!

**Comment by lotri1** [comment]
If you're using the cellNav, you won't be able to select cells text.
If you want key up/down navigation without the cellNav, I wrote a directive based on the cellNav functionality:
        



    function uiGridKeyNav($compile, gridUtil) {

        return {

            require: '^uiGrid',
            scope: false,
            link: function ($scope, $elm, $attrs, uiGridCtrl) {
                var grid = uiGridCtrl.grid;
                var focuser = $compile('<div class="ui-grid-focuser" role="region" aria-live="assertive" aria-atomic="false" tabindex="0" aria-controls="' + grid.id + '-aria-speakable ' + grid.id + '-grid-container' + '" aria-owns="' + grid.id + '-grid-container' + '"></div>')($scope);
                $elm.append(focuser);

                $elm.bind('click', function () {
                    gridUtil.focus.byElement(focuser[0]);
                });

                focuser.bind('keydown', function (e) {
                    $scope.$apply(function () {
                        var selectedEntities,
                            visibleRows,
                            selectedIndex;

                        if (e.keyCode !== 38 && e.keyCode !== 40)
                            return;

                        selectedEntities = grid.api.selection.getSelectedRows();

                        if (selectedEntities.length === 1) {
                            visibleRows = grid.getVisibleRows();

                            selectedIndex = visibleRows.map(function (item) {
                                return item.entity;
                            }).indexOf(selectedEntities[0]);

                            if (selectedIndex < 0)
                                return;

                            if (e.keyCode === 38 && selectedIndex > 0)
                                --selectedIndex;

                            if (e.keyCode === 40 && selectedIndex < grid.options.data.length - 1)
                                ++selectedIndex;

                            grid.api.selection.selectRowByVisibleIndex(selectedIndex);
                            grid.api.core.scrollToIfNecessary(visibleRows[selectedIndex], grid.columns[0]);
                        }
                    });
                });
            }
        };
    }

    var commonDirectivesModule = angular.module('common-directives');

    commonDirectivesModule.directive('uiGridKeyNav', ['$compile', 'gridUtil', uiGridKeyNav]);`

Using the directive on the grid element:

`<div ui-grid="vm.options" ui-grid-selection ui-grid-key-nav>`

**Comment by lentyaishe** [comment]
The best solution that worked for me was updating the original ui-grid.js code and replacing 1 line of code in the following section:
$scope.$on(uiGridCellNavConstants.CELL_NAV_EVENT, function (evt, rowCol, modifierDown) {
            var isFocused = grid.cellNav.focusedCells.some(function(focusedRowCol, index){
                return (focusedRowCol.row === $scope.row && focusedRowCol.col === $scope.col);
            });
            if (isFocused){
              setFocused();
            } else {
              clearFocus();
            }
          });

You'll need to replace:
return (focusedRowCol.row === $scope.row && focusedRowCol.col === $scope.col);
with:
return (focusedRowCol.row === $scope.row);

That's it. The rest is done by the original code.

**Comment by KurtMar** [comment]
I used the directive provided by @lotri1 (Thanks!). I made some modifications for using it with ui.grid.treeView. Specifically added the ability to expand and collapse rows:

`
myAppModule.directive('uiGridKeyNav', ['$compile', 'gridUtil', function($compile, gridUtil) {
    return {

        require: '^uiGrid',
        scope: false,
        link: function ($scope, $elm, $attrs, uiGridCtrl) {
            var grid = uiGridCtrl.grid;
            var focuser = $compile('<div class="ui-grid-focuser" role="region" aria-live="assertive" aria-atomic="false" tabindex="0" aria-controls="' + grid.id + '-aria-speakable ' + grid.id + '-grid-container' + '" aria-owns="' + grid.id + '-grid-container' + '"></div>')($scope);
            $elm.append(focuser);

            $elm.bind('click', function () {
                gridUtil.focus.byElement(focuser[0]);
            });

            focuser.bind('keydown', function (e) {
                $scope.$apply(function () {
                    var selectedEntities,
                        visibleRows,
                        selectedIndex;

                    if (e.keyCode !== 38 && e.keyCode !== 40 && e.keyCode !== 37 && e.keyCode !== 39 && e.keyCode !== 13)
                        return;

                    selectedEntities = grid.api.selection.getSelectedRows();

                    if (selectedEntities.length === 1) {
                        visibleRows = grid.getVisibleRows();

                        selectedIndex = visibleRows.map(function (item) {
                            return item.entity;
                        }).indexOf(selectedEntities[0]);

                        if (selectedIndex < 0)
                            return;
                        // Enter
                        if (e.keyCode === 13) {
                            // You might want to do something here like:
                            //grid.api.selection.unSelectRow(selectedEntities[0], e);
                            //grid.api.selection.selectRow(selectedEntities[0], e);
                            // This way you can pass the enter key to your gridApi.selection.on.rowSelectionChanged event handler
                        // Left
                        } else if (e.keyCode === 37) {
                            grid.api.treeBase.collapseRow(grid.rowHashMap.get(selectedEntities[0]));
                        // Right
                        } else if (e.keyCode === 39) {
                            grid.api.treeBase.expandRow(grid.rowHashMap.get(selectedEntities[0]));
                        // Up / Down
                        } else {
                            if (e.keyCode === 38 && selectedIndex > 0)
                                --selectedIndex;

                            if (e.keyCode === 40 && selectedIndex < grid.options.data.length - 1)
                                ++selectedIndex;
                        
                            grid.api.selection.selectRowByVisibleIndex(selectedIndex, e);
                            grid.api.core.scrollToIfNecessary(visibleRows[selectedIndex], grid.columns[0]);
                        }
                    }
                });
            });
        }
    };
}]);
`

**Comment by stndbye** [comment]
Thanks for the key directive!

Just in case its of any use to anyone, changing $elm.bind like this:
`          
var clickTarget = getEventTarget(e);
if (clickTarget.tagName !== 'INPUT') {
  gridUtil.focus.byElement(focuser[0]);
}

//... where getEventTarget is:
function getEventTarget(e) {
  e = e || window.event;
  return e.target || e.srcElement;
}

`

**Comment by ciceroaferreira** [comment]
@lentyaishe Great solution, it works fine for me. 

Thanks

**Comment by alib192** [comment]
@lentyaishe I really like your directive but it doesn't work for me with the down arrow key, it moves down once but not again. Am I missing something? Also would there be a way to introduce a 'ctrl'/'shift' and arrow option?

Thanks

**Comment by lentyaishe** [comment]
@alib192 Unfortunately I'm quite far now from the project in which I used my workaround. Try to check that you've followed the instructions correctly and didn't miss something since it worked for @ciceroaferreira. Or just try to debug and check why it works only once.

**Comment by sthakker7** [comment]
Any update on this issue? I  believe this bug exists in the latest version of angular ui grid, which is 4.0.7.

**Comment by mportuga** [comment]
Note: add help wanted to this issue to prevent automatic closing
